extensions/ql-vscode/src/qlpack-generator.ts

```diff
@@ -0,0 +1,103 @@
+import { writeFile } from "fs-extra";
+import { dump } from "js-yaml";
+import { join } from "path";
+import { Uri, workspace } from "vscode";
+import { CodeQLCliServer } from "./cli";
+
+export type QueryLanguage =
+  | "csharp"
+  | "cpp"
+  | "go"
+  | "java"
+  | "javascript"
+  | "python"
+  | "ruby"
+  | "swift";
+
+export class QlPackGenerator {
+  private readonly qlpackName: string;
+  private readonly qlpackVersion: string;
+  private readonly header: string;
+  private readonly qlpackFileName: string;
+  private readonly folderUri: Uri;
+
+  constructor(
+    private readonly folderName: string,
+    private readonly queryLanguage: QueryLanguage,
+    private readonly cliServer: CodeQLCliServer,
+    private readonly storagePath: string | undefined,
+  ) {
+    if (this.storagePath === undefined) {
+      throw new Error("Workspace storage path is undefined");
+    }
+    this.qlpackName = `getting-started/codeql-extra-queries-${this.queryLanguage}`;
+    this.qlpackVersion = "1.0.0";
+    this.header = "# This is an automatically generated file.\n\n";
+
+    this.qlpackFileName = "qlpack.yml";
+    this.folderUri = Uri.file(join(this.storagePath, this.folderName));
+  }
+
+  public async generate() {
+    // create QL pack folder and add to workspace
+    await this.createWorkspaceFolder();
+
+    // create qlpack.yml
+    await this.createQlPackYaml();
+
+    // create example.ql
+    await this.createExampleQlFile();
+
+    // create codeql-pack.lock.yml
+    await this.createCodeqlPackLockYaml();
+  }
+
+  private async createWorkspaceFolder() {
+    await workspace.fs.createDirectory(this.folderUri);
+
+    const end = (workspace.workspaceFolders || []).length;
+
+    await workspace.updateWorkspaceFolders(end, 0, {
+      name: this.folderName,
+      uri: this.folderUri,
+    });
+  }
+
+  private async createQlPackYaml() {
+    const qlPackFilePath = join(this.folderUri.fsPath, this.qlpackFileName);
+
+    const qlPackYml = {
+      name: this.qlpackName,
+      version: this.qlpackVersion,
+      dependencies: {
+        [`codeql/${this.queryLanguage}-all`]: "*",
```

Comment:
This line isn't actually necessary since you are calling `packAdd` below, which does the same thing (and more).